internal/pkg/cli/deploy/deploy.go

```diff
@@ -0,0 +1,875 @@
+)
+
+// ImageBuilderPusher builds and pushes an image.
+type ImageBuilderPusher interface {
+	BuildAndPush(docker repository.ContainerLoginBuildPusher, args *dockerengine.BuildArguments) (string, error)
+}
+
+// Uploader uploads a file.
+type Uploader interface {
+	Upload(bucket, key string, data io.Reader) (string, error)
+	ZipAndUpload(bucket, key string, files ...s3.NamedBinary) (string, error)
+}
+
+// Templater stringifies a golang template.
+type Templater interface {
+	Template() (string, error)
+}
+
+// EndpointGetter gets the service discovery endpoint.
+type EndpointGetter interface {
+	ServiceDiscoveryEndpoint() (string, error)
+}
+
+// VersionGetter gets the version.
+type VersionGetter interface {
+	Version() (string, error)
+}
+
+// PublicCIDRBlocksGetter gets the public CIDR blocks.
+type PublicCIDRBlocksGetter interface {
+	PublicCIDRBlocks() ([]string, error)
+}
+
+// CustomResourcesUploader uploads the custom resource files to S3.
+type CustomResourcesUploader interface {
+	UploadEnvironmentCustomResources(upload s3.CompressAndUploadFunc) (map[string]string, error)
+	UploadRequestDrivenWebServiceCustomResources(upload s3.CompressAndUploadFunc) (map[string]string, error)
+	UploadNetworkLoadBalancedWebServiceCustomResources(upload s3.CompressAndUploadFunc) (map[string]string, error)
+}
+
+// SNSTopicsLister lists SNS topics.
+type SNSTopicsLister interface {
+	ListSNSTopics(appName string, envName string) ([]deploy.Topic, error)
+}
+
+// ServiceDeployer uses CloudFormation to deploy a service.
+type ServiceDeployer interface {
+	DeployService(out progress.FileWriter, conf cloudformation.StackConfiguration, bucketName string, opts ...awscloudformation.StackOption) error
+}
+
+// ServiceForceUpdater force updates a service.
+type ServiceForceUpdater interface {
+	ForceUpdateService(app, env, svc string) error
+	LastUpdatedAt(app, env, svc string) (time.Time, error)
+}
+
+// WorkspaceReader reads from the workspace.
+type WorkspaceReader interface {
+	manifestReader
+	pathFinder
+}
+
+// Interpolator interpolates a manifest.
+type Interpolator interface {
+	Interpolate(s string) (string, error)
+}
+
+// progress is the interface to inform the user that a long operation is taking place.
+type Progress interface {
+	Start(label string)
+	Stop(label string)
+}
+
+type manifestReader interface {
+	ReadWorkloadManifest(name string) (workspace.WorkloadManifest, error)
+}
+
+type pathFinder interface {
+	Path() (string, error)
+}
+
+// WorkloadDeployer holds all the metadata and clients needed to deploy a workload.
+type WorkloadDeployer struct {
+	Name     string
+	App      *config.Application
+	Env      *config.Environment
+	ImageTag string
+	S3Bucket string
+
+	// cached variables
+	mft           interface{}
+	workspacePath string
+}
+
+// UploadArtifactsInputs is the input of UploadArtifacts.
+type UploadArtifactsInputs struct {
```

Comment:
To be consistent with the aws sdk go, can we use singular? (example: https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#CreateCapacityProviderInput)

```suggestion
type UploadArtifactsInput struct {
```

Reply:
good idea!